### ISSUE TITLE
Progress tracking

### ISSUE THREAD

**Issue by sinipessala** [jneubert/skos-history#19]
If there are many changes between two KOS versions, the KOS developer needs a way to keep in track which changes she has already seen. This can include plain binary "seen - unseen" or different tags e.g. "unseen - seen - investigate further - problematic - done". Additionally, the tracking can be on triple-level or on concept-level.

How could progress tracking be implemented and which technologies would be useful for it?

**Comment by jneubert** [comment]
Hm - I suppose the tracking can be done only in some kind of workflow system, and it is indedependent from the changes in the vocabulary (and out of scope for here).
However, any workflow system would have to reference the individual changes. That could be done on the level of
1) individual triples - tedious, would involve reification
2) individual skos:Concepts (and perhaps ConceptSchemes, too), as a proxy for all changes which involve that concept. That would structure the workflow by concept rather than by change type, and could cover all kinds of changes (see #18).

@osma I suppose you already have some approach to solve this, and would be interested to learn.

**Comment by osma** [comment]
Yes, we are struggling with this, as we want to be able to replicate most changes in YSA and Allärs to YSO. But YSO is maintained using TopBraid Composer (TBC) which doesn't have a workflow system.

The current approach is to use a skos-history database of YSA and Allärs snapshots, and whenever something has changed in YSA/Allärs between two versions, we insert "something has changed" triples to the relevant YSO concepts, e.g.

```
yso:p16725  yso-update:newPrefLabel  "World Trade Center (New York)"@sv ;
        yso-update:removedPrefLabel  "World Trade Center -- New York"@sv .

yso:p25441  yso-update:newPrefLabel  "kosmoterapia"@fi ;
        yso-update:removedPrefLabel  "kosmomusiikki"@fi .
```

Then in TBC these triples can be found using e.g. SPARQL and the YSO concepts changed - or not, depending on whether the change is relevant - to match what was done in the source thesauri.

I'd love to find a better approach but this is what we have now. If we change to a workflow-enabled maintenance platform for YSO in the future, perhaps we could do this differently. OTOH, the current system where YSA, Allärs and YSO are maintained in parallel is unlikely to last forever, so this can be seen as an interim solution that will eventually be abandoned...

**Comment by jneubert** [comment]
This seems very reasonable in the given environment. It looks somehow similar to what I've briefly investigated with the [concept_history](http://zbw.eu/beta/sparql-lab/?queryRef=https://api.github.com/repos/jneubert/skos-history/contents/sparql/concept_history.rq) query - the properties introduced there are called `sh:prefLabelInsertion`, `sh:prefLabelInsertion` etc. More importantly, the subject is a specific concept delta (e.g. `http://zbw.eu/stw/descriptor/12571-4/version/8.06/delta/8.08`), which is linked to the concept by a `sh:history` property. (Which is unnecessary overhead, if you focus on one set of changes between versions at a time). 
I've not investigated that further, because I hadn't an acutal use case, and because defininig properties for some types of changes seemed a bit arbitrary. For the time beeing, I use a very generic concept-centered approach (as briefly outlined in #18); according queries are linked as "Concept history (RDF/Turtle)" from the STW pages. But the generated format is neither user-friendly to read nor easy to use in applications.
Perhaps it is possible to discover, investigate and propose common patterns. Input from other parties could be very valuable for this.
